keras_nlp/models/gpt2/gpt2_causal_lm.py

```diff
@@ -324,47 +335,109 @@ def next(prompt, state, index):
+                the model. If `True`, `tf.Tensor` outputs will be returned. If
+                `False`, a python list of strings will be return.
+
+        Returns:
+            A string or string list if `preprocessor` is set, and a integer
+            tensor of token IDs if `preprocessor` is `None`.
         """
-        if self.preprocessor is None:
+        input_is_scalar = False
+
+        if isinstance(inputs, tf.data.Dataset) and batch_size is not None:
             raise ValueError(
-                "`self.preprocessor` is `None`, please make sure "
-                "`preprocessor` is set before calling `generate`."
+                "When `inputs` is a `tf.data.Dataset`, please do not provide "
+                "`batch_size` (the dataset is expected to be batched). "
+                f"Received: `batch_size={batch_size}` (of type "
+                f"{type(batch_size)})."
             )
 
-        # Tokenize.
-        prompt = tf.convert_to_tensor(prompt)
-        input_is_scalar = prompt.shape.rank == 0
-        prompt = prompt[tf.newaxis] if input_is_scalar else prompt
-        prompt = self.preprocessor.tokenizer(prompt)
+        if not isinstance(inputs, tf.data.Dataset):
+            if self.preprocessor is not None:
+                inputs = tf.convert_to_tensor(inputs)
+                input_is_scalar = inputs.shape.rank == 0
+                inputs = inputs[tf.newaxis] if input_is_scalar else inputs
+
+            if batch_size is not None:
+                inputs = tf.data.Dataset.from_tensor_slices(inputs)
+                inputs = inputs.batch(batch_size)
+            else:
+                # If `batch_size` is not provided, just wrap in a list to avoid
+                # the overhead of converting to a dataset.
+                inputs = [inputs]
+
+        def preprocess(x):
+            return self.preprocessor(
+                x,
+                sequence_length=max_length,
+                return_labels=False,
+                # We do not append `"<|endoftext|>"` by default, to let the
+                # generation follow directly from the prompt. If working with a
+                # find tuned model where `"<|endoftext|>"` is desired to signal
+                # the end of a prompt, it can be appended to the prompt string.
+                add_end_token=False,
+            )
 
-        # Pad ragged to dense tensors.
-        padded_shape = (None, max_length)
-        min_length = tf.reduce_min(prompt.row_lengths())
-        input_mask = tf.ones_like(prompt, tf.bool).to_tensor(shape=padded_shape)
-        prompt = prompt.to_tensor(shape=padded_shape)
+        if self.preprocessor is not None:
+            if isinstance(inputs, tf.data.Dataset):
+                inputs = inputs.map(preprocess, tf.data.AUTOTUNE)
+                inputs = inputs.prefetch(tf.data.AUTOTUNE)
+            else:
+                inputs = [preprocess(batch) for batch in inputs]
```

Comment:
`preprocess` calls `self.preprocessor`, how does this work if `self.preprocessor is None`?

Reply:
Note that this is inside a branch `if self.preprocessor is not None:`.